rust/connlib/tunnel/src/peer.rs

```diff
@@ -335,6 +535,87 @@ pub struct ClientOnGateway {
     ipv6: Ipv6Addr,
     resources: HashMap<ResourceId, Vec<ResourceOnGateway>>,
     filters: IpNetworkTable<FilterEngine>,
+    permanent_translations: HashMap<IpAddr, (IpAddr, ResourceId, DomainName, Instant, bool)>,
+    nat_table: NatTable,
+    buffered_events: VecDeque<GatewayEvent>,
+}
+
+#[derive(Default, Debug)]
+struct NatTable {
+    table: BiMap<(Protocol, IpAddr), (Protocol, IpAddr)>,
+    last_seen: HashMap<(Protocol, IpAddr), Instant>,
+}
+
+impl NatTable {
+    fn expire(&mut self, now: Instant) {
+        let mut removed = Vec::new();
+        for (r, e) in self.last_seen.iter() {
+            if now.duration_since(*e) >= Duration::from_secs(60) {
+                self.table.remove_by_right(r);
+                removed.push(*r);
+            }
+        }
+
+        for r in removed {
+            self.last_seen.remove(&r);
+        }
+    }
+
+    fn translate(
+        &mut self,
+        outgoing_pkt: &IpPacket,
+        real_address: IpAddr,
+        now: Instant,
+    ) -> Option<(Protocol, IpAddr)> {
+        let source_protocol = outgoing_pkt.source_protocol()?;
+        if let Some(res) = self
+            .table
+            .get_by_left(&(source_protocol, outgoing_pkt.destination()))
+        {
+            if res.1 == real_address {
+                self.last_seen.insert(*res, now);
+                return Some(*res);
+            }
```

Comment:
When can the case happen that we have a mapping but the `real_address` doesn't match? Should we log something here?

Reply:
when the DNS was refreshed